File: code/functions/cube_root.py
# cube_root.py
"""
Obtain cube root using Newton's method
"""


def cube_root(x):
    def improve(guess):
        return((x/guess**2 + 2*guess)/3)

    def good_enough(guess):
        return(improve(guess) == guess)

    def cube_iter(guess):
        if good_enough(guess):
            return guess
        else:
            return cube_iter(improve(guess))

    return cube_iter(1.0)

File: code/functions/cuberoot.R
cuberoot <- function(x) {
    improve <- function(guess) {
        (x/guess^2 + 2*guess)/3
        }
    goodenough <- function(guess) {
        improve(guess) == guess
    }
    cube_iter <- function(guess) {
        ifelse(goodenough(guess),
               guess,
               cube_iter(improve(guess)))
    }
    return(cube_iter(1.0))
}

    
